horovod/spark/lightning/remote.py

```diff
@@ -121,6 +121,12 @@ def train(serialized_model):
             _val_steps_per_epoch = val_steps_per_epoch if val_steps_per_epoch else 1.0
 
             cuda_available = torch.cuda.is_available()
+            # We need to check all ranks have same device type for traning.
+            # Horovod doesn't support heterogeneous allreduce for gradients.
+            cuda_avail_list = hvd.allgather_object(cuda_available, name='device type')
+            if hvd.rank() == 0:
+                assert cuda_avail_list.count(cuda_available) == hvd.size(), "All ranks don't have same device type!"
```

Comment:
Wouldn't you want to assert this on all ranks, not just rank 0?

Reply:
That would be the same result, since you have all nodes status in a `cuda_avail_list`.
We are doing a count op here so it covers all cases:
1. all False/True: assert passes
2. The `cuda_avail_list` has both True/False inside: `cuda_avail_list.count(cuda_available) == hvd.size()` will fail no matter `cuda_available` is True or False.

In other words, allgather returns a global result, which rank checks `cuda_avail_list.count(cuda_available) == hvd.size()`doesn't matter.

Reply:
Right, but because you're only checking this on rank 0, you're relying on the fact that rank.0 will fail, and then some time in the future the other ranks will attempt to perform a collective op (like `hvd.allreduce`) which will cause the other ranks to fail as well. If you perform the check on all ranks, then they all fail at the same time, which is more reliable and fails faster. Does that make sense?

Reply:
I see.

Reply:
fixed